### DISCUSSION TITLE
Screen flicker

### DISCUSSION THREAD

**Discussion by grakoczy** [picocomputer/community#126] [Q&A] [answered]
Hi,

Is there any way to avoid screen flickering when I cycle draw lines / erase screen? Both operations take a while and it's visible as flicker.
I've never really programmed 8 bit machines back in the 80's so I don't know the tricks. 
There is are no screen tearing so I think it's not a vsync issue.
Here is how it looks so far.

https://github.com/user-attachments/assets/50af9d2b-7fd1-43a0-a33b-0f1c69dc3077

**Comment by rumbledethumps** [comment]
Double buffering.

**Comment by grakoczy** [comment]
Is this possible on RP6502?

**Reply by WojciechGw** [reply]
320x240 yes
640x480 no  (due to XRAM size)

proof below
(blinking pixels in upper corners indicate buffer switching)

https://github.com/user-attachments/assets/74567f03-d7de-4f89-9201-e8afe3263482

**Reply by grakoczy** [reply]
Hi,
Looks great. How did you do that?

**Reply by WojciechGw** [reply, marked as answer]
First of all thank You very much for 3d transformation functions :)

Magic is done by drawing things to two different address spaces in XRAM and switching between them

0. clear inactive buffer
1. draw things to inactive buffer
2. switch to inactive buffer
3. goto 0.

I've made some changes to tonyvr's bitmap_graphics library and Your main.c
so all functions are double buffering friendly now

in bitmap_graphics_db.c
```
void switch_buffer(uint16_t buffer_data_address)
{
    xram0_struct_set(canvas_struct, vga_mode3_config_t, xram_data_ptr, buffer_data_address);
}
```

main.c in main loop
```
[...]
     buffers[0] = 0x0000;
     buffers[1] = 0x2580;
[...]
      // draw on inactive buffer
      erase_buffer(buffers[!active_buffer]);
      draw_cube(angleX, angleY, angleZ, WHITE, mode, buffers[!active_buffer]);
      switch_buffer(buffers[!active_buffer]);
      active_buffer = !active_buffer;
[...]
```

in main CMakeList.txt
```
add_executable(cube)
rp6502_executable(cube)
target_sources(cube PRIVATE
    src/colors.c
    src/bitmap_graphics_db.c
    src/main.c
)
```
all changed files in .zip
[3dcube_DoubleBuffering.zip](https://github.com/user-attachments/files/17351423/3dcube_DoubleBuffering.zip)

Bon apetite !

**Reply by grakoczy** [reply]
Thanks! 
Now it works great :)

--
Grzegorz

**Reply by WojciechGw** [reply]
my pleasure :)

double buffering capable resolutions

canvas_w | canvas_h | bpp | pixels | buffer size
-- | -- | -- | -- | --
640 | 360 | 1 | 230400 | 28800
320 | 240 | 2 | 76800 | 19200
320 | 240 | 1 | 76800 | 9600
320 | 180 | 4 | 57600 | 28800
320 | 180 | 2 | 57600 | 14400
320 | 180 | 1 | 57600 | 7200
240 | 124 | 8 | 29760 | 29760
240 | 124 | 4 | 29760 | 14880
240 | 124 | 2 | 29760 | 7440
240 | 124 | 1 | 29760 | 3720

sorry, just noticed garbage in main.c
You can delete
```
int16_t cube_vertices_c[8][3] = {
    {-4096, -4096, -4096}, {4096, -4096, -4096}, {4096, 4096, -4096}, {-4096, 4096, -4096},  // Back face
    {-4096, -4096,  4096}, {4096, -4096,  4096}, {4096, 4096,  4096}, {-4096, 4096,  4096}   // Front face
};
```
it was for testing purposes only

**Comment by WojciechGw** [comment]
continuing the topic,
take a look at this

[3dcube.zip](https://github.com/user-attachments/files/17590808/3dcube.zip)

[source[3dcube].zip](https://github.com/user-attachments/files/17590810/source.3dcube.zip)

**Comment by grakoczy** [comment]
Nice! 
Merged to the repository: [https://github.com/grakoczy/RP6502-3D-cube-llvm-mos](url)

**Reply by vruumllc** [reply]
Great demo! I love it.
